### PR TITLE
`symmetric_matrix_rank_1_update`: kokkos impl

### PR DESCRIPTION
**Pull request by MikolajZuzek**
This is Kokkos implementation of `symmetric_matrix_rank_1_update`.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@fnrizzi has been touching a lot of this code of late, but it looks like you know that : - )  I'll wait until you two say it's ready to merge.

**Comment by fnrizzi** [comment]
@mhoemmen yes, I have a plan with @MikolajZuzek so he is doing all the *_update functions :) so you can safely review this since it does not interfere with anything i am doing!

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:44]: Please don't declare this `static`, as it prevents `is_int_mul_safe` from being `constexpr`.
```suggestion
    constexpr auto max_val = std::numeric_limits<decltype(a * b)>::max();
```

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:86]: Would you consider instead defining `triangle_layout_match`?  It's harder for readers to do the double negative in their heads.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:129]: Would you consider encapsulating the logger instead of using `<iostream>` objects directly?  For example:

```c++
extern void logKokkos(std::string_view functionName);
```

and then
```c++
#if defined(KOKKOS_STDBLAS_ENABLE_TESTS)
logKokkos("symmetric_matrix_rank1_update");
#endif
```

**Review by mhoemmen — CHANGES_REQUESTED** [review]
The bug in `is_int_mul_safe` is a "must change"; other comments are suggestions.  Thanks!

**Comment by mhoemmen** [comment]
@MikolajZuzek Thanks for working on this, btw! : - )

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:44]: Fixed on https://github.com/kokkos/stdBLAS/pull/167/commits/4e301eee45d80de3949c091a33dba4e38437ed79

File: tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp
Comment (MikolajZuzek) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:56]: Would this be better: https://github.com/kokkos/stdBLAS/commit/ab0f592b3cd5e41ac02570fd2b5cdb50cbadd30d ? (it's on a "preview" branch, not yet included in this PR)

I can't get `mdspan`'s CTAD working for const values nor for rank 2 (like in [Wiki](https://github.com/kokkos/mdspan/wiki/A-Gentle-Introduction-to-mdspan#what-is-mdspan)).

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/878fa923c8683602c949548fb6e5c9f666ecf622

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thank you, fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/034ea2a445d24be0f92c604146b1e6008017f2d4.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: I'm sorry for trying to sneak that in... I removed `is_int_mul_safe()` on https://github.com/kokkos/stdBLAS/pull/167/commits/c209631bc02ba6038d105f69d6fbd2fe48004e13, because it's no longer needed here*, but your your corrected version is preserved for future use: thank you very much for super useful lesson on `constexpr`.

*I thought of fine grained parallelism with one matrix element per thread (instead of whole row/column) - which requires `extent0 * extent1` not to overflow, thus `is_int_mul_safe()` - but then realized it doesn't make any sense on CPU because of extra per-element cost on index splitting (`i = ij / extent0` and it's worse for triangles), so I dropped that route as premature optimization.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thank you, fix explained [above](https://github.com/kokkos/stdBLAS/pull/167#discussion_r849897213).

File: tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp
Comment (mhoemmen) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:56]: @MikolajZuzek Thanks for trying this out!  Some mdspan changes for better CTAD support may be recent or still in flight in the reference implementation.  These are test-only functions, so I wouldn't worry too much.

File: tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp
Comment (mhoemmen) [inline tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:179]: Does this actually work on the GPU?  It's using `std::complex`.

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:159]:
```suggestion
  // P1673 constraints (redundant to mdspan extents in the header)
```

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:112]:
```suggestion
// Note: will only signal failure for layout_blas_packed with different triangle
```

**Review by mhoemmen — APPROVED** [review]
Thanks for the fixes!  I'm OK merging this, but there are a couple issues remaining:

1. some use of `std::complex` in a `KOKKOS_LAMBDA`; not sure if OK; and
2. a couple spelling errors in comments.

If you're in a hurry, I'll be happy to merge this, else I'll wait for fixes.  Thanks!

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thanks, fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/9da170bac360e3d9f05ff7e7732fdddf475af4e3

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp]: Thanks, fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/9da170bac360e3d9f05ff7e7732fdddf475af4e3

**Comment by MikolajZuzek** [comment]
> If you're in a hurry, I'll be happy to merge this, else I'll wait for fixes. Thanks!

@mhoemmen Thank you for your thorough reviews and great comments! No hurry on my part: it's better to find and fix as many things as we can, so they don't get duplicated on other PRs.

File: tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp
Comment (MikolajZuzek) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:179]: I tried building for CUDA but failed with `nvcc` not being able to compile `mdspan` (the errors didn't look fatal, but I was not sure if it's worth digging into).
Would you know if `stdBLAS` should build for CUDA regardless of requiring C++20 while [`nvcc` only supports C++17](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#options-for-altering-compiler-linker-behavior-std) ?

Comment (MikolajZuzek) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:179]: Oh, maybe it was about `nvcc` complaining `std::complex` - I'll check that.

Comment (MikolajZuzek) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:179]: Probably it's also not a good idea to use parallel elements in tests -
or do it through `ParallelMatrixVisitor` with `is_int_mul_safe` (see the discussion [above](https://github.com/kokkos/stdBLAS/pull/167#discussion_r846204678))

Comment (MikolajZuzek) [reply tests/kokkos-based/symmetric_matrix_rank1_update_kokkos.cpp:179]: Fixed with https://github.com/kokkos/stdBLAS/pull/167/commits/83d174227fd8cd9aebf11c76e14646401f713156: we cast `mdspan`s to `Kokkos::View`s correctly now, so inside lambda we'll have `std::complex` replaced with `Kokkos::complex` by `mdspan_to_view()`.

@mhoemmen thank you for pointing this out!

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (mhoemmen) [inline tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:147]: See https://github.com/kokkos/stdBLAS/issues/173 .

**Review by mhoemmen — APPROVED** [review]
Thanks for the contribution!

File: tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp
Comment (MikolajZuzek) [reply tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_1_update.hpp:147]: Addressed on #179.

**Pull request merged**
